tests/Propel/Tests/Generator/Builder/Om/QueryBuilderTest.php

```diff
@@ -1389,4 +1393,82 @@ public function testPruneCompositeKey()
         $nbBookListRel = BookListRelQuery::create()->prune($testBookListRel)->count();
         $this->assertEquals(1, $nbBookListRel, 'prune() removes an object from the result');
     }
+
+    /**
+     * @return void
+     */
+    public function testFindPkSimpleIsAbstractWhenTableIsAbstract(): void
+    {
+        $databaseXml = '
+<database>
+    <table name="my_table" abstract="true">
+        <column name="id" type="integer" primaryKey="true" />
+    </table>
+</database>
+';
+        $script = TestableQueryBuilder::forTableFromXml($databaseXml, 'my_table')->buildScript('addFindPkSimple');
+        $expectedMethodHead = '    abstract protected function findPkSimple($key, ConnectionInterface $con);';
+        $msg = 'Query class for abstract table should have abstract findPkSimple() method';
+        $this->assertStringContainsString($expectedMethodHead, $script, $msg);
+    }
+
+    /**
+     * @return void
+     */
+    public function testFindPkSimpleIsNotAbstractWhenTableIsAbstractWithInheritance(): void
+    {
+        $databaseXml = '
+<database>
+    <table name="my_table" abstract="true">
+        <column name="id" type="integer" primaryKey="true" />
+        <column name="class_key" type="integer" inheritance="single">
+            <inheritance key="1" class="class1" />
+            <inheritance key="2" class="class2" extends="my_table" />
+        </column>
+    </table>
+</database>
+';
+        $script = TestableQueryBuilder::forTableFromXml($databaseXml, 'my_table')->buildScript('addFindPkSimple');
+        $expectedMethodHead = '    protected function findPkSimple($key, ConnectionInterface $con)';
+        $msg = 'Query class for abstract table should not have abstract findPkSimple() method if table uses inheritance';
+        $this->assertStringContainsString($expectedMethodHead, $script, $msg);
+    }
+}
+
+class TestableQueryBuilder extends QueryBuilder
+{
+    /**
+     * Build a TestableQueryBuilder for a table given in schema xml format.
+     *
+     * @param string $schemaXml
+     * @param string $tableName
+     *
+     * @return self
+     */
+    public static function forTableFromXml(string $schemaXml, string $tableName): self
```

Comment:
self and static (return) types should be removed. They should be documented only.

Reply:
Oh, the code sniffer actually changed the return type from `TestableQueryBuilder` to `self`, and now does not complain anymore. Do you still want me to change it?

I moved the class into its own file. While doing so, I found that there already is a file that collects other classes used in the test file (`QueryBuilderTestClasses.php`). I think I have seen this pattern in other tests as well. The folder has 40 files now, which makes it all too easy to overlook something like this.
Note that autoloading classes that extend Bookstore will cause CI to fail during agnostic tests. I think it is because the fixtures are not generated. Those files have to be included manually. So autoloading is not a general solution, or needs at least needs an update of how CI works.